File: test/yosupo-shortest-path.test.cpp
#define PROBLEM "https://judge.yosupo.jp/problem/shortest_path"

#include "../cpp/graph.hpp"

int main() {
    int n, m, s, t; std::cin >> n >> m >> s >> t;
    Graph<long long> g(n); g.read(m, 0, true, true);
    auto [dist, prev] = g.shortest_path(s);
    if(dist[t] == -1) {
        std::cout << -1 << '\n';
        return 0;
    }
    std::vector<std::pair<int, int>> route;
    int cur = t;
    while(cur != s) {
        route.emplace_back(prev[cur].src, prev[cur].dst);
        cur = prev[cur].src;
    }
    std::cout << dist[t] << ' ' << route.size() << '\n';
    for(auto it = route.rbegin(); it != route.rend(); it++) {
        std::cout << it->first << ' ' << it->second << '\n';
    }
}

File: test/yosupo-shortest-path.test.py
# verification-helper: PROBLEM https://judge.yosupo.jp/problem/shortest_path
from py.graph import Graph


def main():
    n, m, s, t = map(int, input().split())
    g = Graph(n)
    g.read(m, 0, True, True)
    dist, prev = g.shortest_path(s, True)
    if dist[t] == -1:
        exit(print(-1))
    route = []
    cur = t
    while cur != s:
        route.append((prev[cur].src, prev[cur].dst))
        cur = prev[cur].src
    print(dist[t], len(route))
    for src, dst in route[::-1]:
        print(src, dst)


if __name__ == "__main__":
    main()
